src/components/ZoomRect/index.js

```diff
@@ -0,0 +1,367 @@
+        }
+      });
+    });
+    this.lastTouch = { ...newTouchPosition };
+    return updates;
+  };
+
+  performTouchZoom = (touches, distances) => {
+    const { itemIds, subDomainsByItemId, zoomAxes, width, height } = this.props;
+    const [touchOne, touchTwo] = touches;
+    const { x: touchOneX, y: touchOneY } = this.getOffset(touchOne);
+    const { x: touchTwoX, y: touchTwoY } = this.getOffset(touchTwo);
+    const centers = {
+      [Axes.time]: (touchOneX + touchTwoX) / 2,
+      [Axes.x]: (touchOneX + touchTwoX) / 2,
+      [Axes.y]: (touchOneY + touchTwoY) / 2,
+    };
+    const deltas = {
+      [Axes.time]: Math.abs(touchOneX - touchTwoX),
+      [Axes.x]: Math.abs(touchOneX - touchTwoX),
+      [Axes.y]: Math.abs(touchOneY - touchTwoY),
+    };
+    const multipliers = {
+      [Axes.time]: -1,
+      [Axes.x]: -1,
+      [Axes.y]: 1,
+    };
+    // This is almost the same as the `distances` object, except the height is
+    // un-inverted for y axis.
+    const measurements = {
+      [Axes.time]: width,
+      [Axes.x]: width,
+      [Axes.y]: height,
+    };
+
+    const updates = {};
+    itemIds.forEach(itemId => {
+      updates[itemId] = {};
+      Axes.ALL.filter(axis => zoomAxes[axis] && this.lastDeltas[axis]).forEach(
+        axis => {
+          const subDomain = (subDomainsByItemId[itemId] || {})[axis];
+          const subDomainRange = subDomain[1] - subDomain[0];
+          const percentFromEnd = centers[axis] / measurements[axis];
+
+          const zoomFactor =
+            multipliers[axis] *
+            ((deltas[axis] - this.lastDeltas[axis]) / distances[axis]);
+
+          // Figure out the value on the scale where the mouse is so that the
+          // new subdomain does not shift.
+          const valueAtCenter = subDomain[0] + subDomainRange * percentFromEnd;
+
+          // How big the next subdomain is going to be
+          const newSpan = subDomainRange * (1 + zoomFactor);
+
+          // Finally, place this new span into the subdomain, centered about the
+          // mouse, and correctly (proportionately) split above & below so that
+          // theaxis is stable.
+          updates[itemId][axis] = [
+            valueAtCenter - newSpan * percentFromEnd,
+            valueAtCenter + newSpan * (1 - percentFromEnd),
+          ];
+        }
+      );
+    });
+    this.lastDeltas = { ...deltas };
+    return updates;
+  };
+
+  syncZoomingState = () => {
+    const { zoomAxes } = this.props;
+    if (Object.keys(zoomAxes).find(axis => zoomAxes[axis])) {
+      this.rectSelection.on('touchend', this.onTouchEnd, true);
+      this.rectSelection.on('touchmove', this.onTouchMove, true);
+      this.rectSelection.on('touchstart', this.onTouchStart, true);
+      this.rectSelection.call(this.zoom.on('zoom', this.zoomed));
+      this.rectSelection.on('dblclick.zoom', null);
+    } else {
+      this.rectSelection.on('.zoom', null);
+    }
+  };
+
+  zoomed = () => {
```

Comment:
is `zoomed` used for desktop zoom whereas `performTouchZoom` is used for mobile? If this is the case can you document that?

Reply:
Sure.